openssl/src/ssl/callbacks.rs

```diff
@@ -416,3 +424,92 @@ where
         callback(ssl, slice) as c_int
     }
 }
+
+#[cfg(all(feature = "v111", ossl111))]
+pub struct CustomExtAddState(Option<Cow<'static, [u8]>>);
+
+#[cfg(all(feature = "v111", ossl111))]
+pub extern "C" fn raw_custom_ext_add<F>(ssl: *mut ffi::SSL, _: c_uint,
+                                        context: c_uint,
+                                        out: *mut *const c_uchar,
+                                        outlen: *mut size_t, x: *mut ffi::X509,
+                                        chainidx: size_t, al: *mut c_int,
+                                        _: *mut c_void)
+                                        -> c_int
+    where F: Fn(&mut SslRef, ExtensionContext, Option<(usize, &X509Ref)>) -> Result<Option<Cow<'static, [u8]>>, SslAlert> + 'static
+{
+    unsafe {
+        let ssl_ctx = ffi::SSL_get_SSL_CTX(ssl as *const _);
+        let callback = ffi::SSL_CTX_get_ex_data(ssl_ctx, get_callback_idx::<F>());
+        let callback = &*(callback as *mut F);
+        let state = {
+            let idx = get_ssl_callback_idx::<CustomExtAddState>();
+            let ptr = ffi::SSL_get_ex_data(ssl, idx);
+            if ptr.is_null() {
+                let x = Box::into_raw(Box::<CustomExtAddState>::new(CustomExtAddState(None))) as *mut c_void;
+                ffi::SSL_set_ex_data(ssl, idx, x);
+                x
+            } else {
+                ptr
+            }
+        };
+        let state = &mut (*(state as *mut CustomExtAddState)).0;
+        let ssl = SslRef::from_ptr_mut(ssl);
+        let ectx = ExtensionContext::from_bits(context).unwrap();
```

Comment:
This will panic if openssl adds any new fields to this - make it via `ExtensionContext { bits: context }` instead to avoid that.